parquet-column/src/main/java/org/apache/parquet/io/MessageColumnIO.java

```diff
@@ -294,17 +301,33 @@ private void writeNullForMissingFieldsAtCurrentLevel() {
 
     private void writeNull(ColumnIO undefinedField, int r, int d) {
       if (undefinedField.getType().isPrimitive()) {
-        columnWriter[((PrimitiveColumnIO)undefinedField).getId()].writeNull(r, d);
+        columnWriter[((PrimitiveColumnIO) undefinedField).getId()].writeNull(r, d);
       } else {
-        GroupColumnIO groupColumnIO = (GroupColumnIO)undefinedField;
-        writeNullToLeaves(groupColumnIO, r, d);
+        GroupColumnIO groupColumnIO = (GroupColumnIO) undefinedField;
+        // only cache the repetition level, the definition level should always be the definition level of the parent node
+        cacheNullForGroup(groupColumnIO, r);
       }
     }
 
-    private void writeNullToLeaves(GroupColumnIO group, int r, int d) {
-      for(ColumnWriter leafWriter: groupToLeafWriter.get(group)) {
-        leafWriter.writeNull(r,d);
+    private void cacheNullForGroup(GroupColumnIO group, int r) {
+      if (!groupNullCache.containsKey(group)) {
+        groupNullCache.put(group, new IntArrayList());
       }
+      groupNullCache.get(group).add(r);
+    }
+
+    private void writeNullToLeaves(GroupColumnIO group) {
+      IntArrayList nullCache = groupNullCache.get(group);
+      if (nullCache == null || nullCache.isEmpty())
+        return;
+
+      int parentDefinitionLevel = group.getParent().getDefinitionLevel();
+      for (ColumnWriter leafWriter : groupToLeafWriter.get(group)) {
+        for (int repetitionLevel : groupNullCache.get(group)) {
```

Comment:
the `for x : y` syntax works for:
a) Arrays
b) `java.lang.Iterable`s

IntArrayList *is* an `Iterable<Integer>`, but that's sort of a convenience that will bypass all the care that went into avoiding boxing / unboxing in IntArrayList to start with.

Looks like what you want is the `intIterator` method which will return you a IntIterator: http://fastutil.di.unimi.it/docs/it/unimi/dsi/fastutil/ints/IntIterator.html

so something like this:

```
for (IntIterator iter = nullCache.intIterator(); iter.hasNext();) {
  int repetitionLevel = iter.nextInt();
  leafWriter.writeNull(repetitionLevel, parentDefinitionLevel);
}
```